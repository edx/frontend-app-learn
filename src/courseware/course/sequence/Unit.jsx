import React, {
  Suspense,
  useEffect,
  useRef,
  useState,
  useLayoutEffect,
} from 'react';
import { useDispatch } from 'react-redux';
import PropTypes from 'prop-types';
import { getConfig } from '@edx/frontend-platform';
import { injectIntl, intlShape } from '@edx/frontend-platform/i18n';
import { Modal } from '@edx/paragon';
import messages from './messages';
import BookmarkButton from '../bookmark/BookmarkButton';
import { useModel } from '../../../generic/model-store';
import PageLoading from '../../../generic/PageLoading';
import { processEvent } from '../../../course-home/data/thunks';
import { fetchCourse } from '../../data/thunks';

const LockPaywall = React.lazy(() => import('./lock-paywall'));
const LockPaywallValuePropExperiment = React.lazy(() => import('./lock-paywall-value-prop'));

/**
 * We discovered an error in Firefox where - upon iframe load - React would cease to call any
 * useEffect hooks until the user interacts with the page again.  This is particularly confusing
 * when navigating between sequences, as the UI partially updates leaving the user in a nebulous
 * state.
 *
 * We were able to solve this error by using a layout effect to update some component state, which
 * executes synchronously on render.  Somehow this forces React to continue it's lifecycle
 * immediately, rather than waiting for user interaction.  This layout effect could be anywhere in
 * the parent tree, as far as we can tell - we chose to add a conspicuously 'load bearing' (that's
 * a joke) one here so it wouldn't be accidentally removed elsewhere.
 *
 * If we remove this hook when one of these happens:
 * 1. React figures out that there's an issue here and fixes a bug.
 * 2. We cease to use an iframe for unit rendering.
 * 3. Firefox figures out that there's an issue in their iframe loading and fixes a bug.
 * 4. We stop supporting Firefox.
 * 5. An enterprising engineer decides to create a repo that reproduces the problem, submits it to
 *    Firefox/React for review, and they kindly help us figure out what in the world is happening
 *    so  we can fix it.
 *
 * This hook depends on the unit id just to make sure it re-evaluates whenever the ID changes.  If
 * we change whether or not the Unit component is re-mounted when the unit ID changes, this may
 * become important, as this hook will otherwise only evaluate the useLayoutEffect once.
 */
function useLoadBearingHook(id) {
  const setValue = useState(0)[1];
  useLayoutEffect(() => {
    setValue(currentValue => currentValue + 1);
  }, [id]);
}

function Unit({
  courseId,
  format,
  onLoaded,
  id,
  intl,
}) {
  let iframeUrl = `${getConfig().LMS_BASE_URL}/xblock/${id}?show_title=0&show_bookmark_button=0`;
  if (format) {
    iframeUrl += `&format=${format}`;
  }

  const [iframeHeight, setIframeHeight] = useState(0);
  const [hasLoaded, setHasLoaded] = useState(false);
  const [modalOptions, setModalOptions] = useState({ open: false });
  const [valuePropExperimentLock, setValuePropExperimentLock] = useState(window.valuePropExperimentLock);
  window.toggleValuePropPaywallLock = () => {
    window.valuePropExperimentLock = !valuePropExperimentLock;
    setValuePropExperimentLock(!valuePropExperimentLock);
  };

  const unit = useModel('units', id);
  const course = useModel('courses', courseId);
  const {
    contentTypeGatingEnabled,
  } = course;

  const dispatch = useDispatch();

  // Do not remove this hook.  See function description.
  useLoadBearingHook(id);

  // We use this ref so that we can hold a reference to the currently active event listener.
  const messageEventListenerRef = useRef(null);
  useEffect(() => {
    function receiveMessage(event) {
      const { type, payload } = event.data;
      if (type === 'plugin.resize') {
        setIframeHeight(payload.height);
        if (!hasLoaded && iframeHeight === 0 && payload.height > 0) {
          setHasLoaded(true);
          if (onLoaded) {
            onLoaded();
          }
        }
      } else if (type === 'plugin.modal') {
        payload.open = true;
        setModalOptions(payload);
      }
    }
    // If we currently have an event listener, remove it.
    if (messageEventListenerRef.current !== null) {
      global.removeEventListener('message', messageEventListenerRef.current);
      messageEventListenerRef.current = null;
    }
    // Now add our new receiveMessage handler as the event listener.
    global.addEventListener('message', receiveMessage);
    // And then save it to our ref for next time.
    messageEventListenerRef.current = receiveMessage;
    // When the component finally unmounts, use the ref to remove the correct handler.
    return () => global.removeEventListener('message', messageEventListenerRef.current);
  }, [id, setIframeHeight, hasLoaded, iframeHeight, setHasLoaded, onLoaded]);

  return (
    <div className="unit">
      <h2 className="mb-0 h4">{unit.title}</h2>
      <BookmarkButton
        unitId={unit.id}
        isBookmarked={unit.bookmarked}
        isProcessing={unit.bookmarkedUpdateState === 'loading'}
      />
      { contentTypeGatingEnabled && unit.graded && (
        <Suspense
          fallback={(
            <PageLoading
              srMessage={intl.formatMessage(messages['learn.loading.content.lock'])}
            />
          )}
        >
          {(valuePropExperimentLock)
            ? <LockPaywallValuePropExperiment courseId={courseId} />
            : <LockPaywall courseId={courseId} />}
        </Suspense>
      )}
      {!hasLoaded && (
        <PageLoading
          srMessage={intl.formatMessage(messages['learn.loading.learning.sequence'])}
        />
      )}
      {modalOptions.open && (
        <Modal
          body={(
            <>
              {modalOptions.body
                ? <div className="unit-modal">{ modalOptions.body }</div>
                : (
                  <iframe
                    title={modalOptions.title}
                    allow="microphone *; camera *; midi *; geolocation *; encrypted-media *"
                    frameBorder="0"
                    src={modalOptions.url}
                    style={{
                      width: '100%',
                      height: '100%',
                    }}
                  />
                )}
            </>
          )}
          onClose={() => { setModalOptions({ open: false }); }}
          open
          dialogClassName="modal-lti"
        />
      )}
      <div className="unit-iframe-wrapper">
        <iframe
          id="unit-iframe"
          title={unit.title}
          src={iframeUrl}
          allowFullScreen
          height={iframeHeight}
          scrolling="no"
          referrerPolicy="origin"
          onLoad={() => {
            window.onmessage = function handleResetDates(e) {
              if (e.data.event_name) {
                dispatch(processEvent(e.data, fetchCourse));
              }
            };
          }}
        />
      </div>
    </div>
  );
}

Unit.propTypes = {
  courseId: PropTypes.string.isRequired,
  format: PropTypes.string,
  id: PropTypes.string.isRequired,
  intl: intlShape.isRequired,
  onLoaded: PropTypes.func,
};

Unit.defaultProps = {
  format: null,
  onLoaded: undefined,
};

export default injectIntl(Unit);
