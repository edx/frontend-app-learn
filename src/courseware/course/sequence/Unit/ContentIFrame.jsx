import PropTypes from 'prop-types';
import React, { useEffect, useState } from 'react';

import { ErrorPage } from '@edx/frontend-platform/react';
import { StrictDict } from '@edx/react-unit-test-utils';
import {
  ModalDialog, Modal, Icon, PageBanner,
} from '@openedx/paragon';

import PageLoading from '@src/generic/PageLoading';
import { getAuthenticatedHttpClient } from '@edx/frontend-platform/auth';
import { WarningFilled } from '@openedx/paragon/icons';
import * as hooks from './hooks';

/**
 * Feature policy for iframe, allowing access to certain courseware-related media.
 *
 * We must use the wildcard (*) origin for each feature, as courseware content
 * may be embedded in external iframes. Notably, xblock-lti-consumer is a popular
 * block that iframes external course content.

 * This policy was selected in conference with the edX Security Working Group.
 * Changes to it should be vetted by them (security@edx.org).
 */
export const IFRAME_FEATURE_POLICY = (
  'microphone *; camera *; midi *; geolocation *; encrypted-media *, clipboard-write *'
);

export const testIDs = StrictDict({
  contentIFrame: 'content-iframe-test-id',
  modalIFrame: 'modal-iframe-test-id',
});

const IFrameComponent = ({
  url, shouldShowContent, hasLoaded, loadingMessage, showError, title, contentIFrameProps,
}) => {
  const [httpStatusCode, setHttpStatusCode] = useState(500);
  useEffect(() => {
    getAuthenticatedHttpClient().get(url).then(res => {
      setHttpStatusCode(res.status);
    }).catch((error) => {
      setHttpStatusCode(error.response.status);
    });
  }, []);
  return (
    <>
      {(shouldShowContent && !hasLoaded) && (
      // eslint-disable-next-line no-nested-ternary
        showError ? (httpStatusCode >= 400 && httpStatusCode < 500
          ? (
            <PageBanner
              variant="warning"
            >
              <Icon src={WarningFilled} className="mie-2" /> Please signin to view this content
            </PageBanner>
          ) : <ErrorPage />) : <PageLoading srMessage={loadingMessage} />
      )}
      {shouldShowContent && (
      <div className="unit-iframe-wrapper">
        <iframe title={title} {...contentIFrameProps} data-testid={testIDs.contentIFrame} />
      </div>
      )}
    </>
  );
};
IFrameComponent.propTypes = {
  url: PropTypes.string.isRequired,
  shouldShowContent: PropTypes.bool.isRequired,
  hasLoaded: PropTypes.bool.isRequired,
  loadingMessage: PropTypes.bool.isRequired,
  showError: PropTypes.bool.isRequired,
  title: PropTypes.string.isRequired,
  // eslint-disable-next-line react/forbid-prop-types
  contentIFrameProps: PropTypes.any.isRequired,
};

const ContentIFrame = ({
  iframeUrl,
  shouldShowContent,
  loadingMessage,
  id,
  elementId,
  onLoaded,
  title,
}) => {
  const {
    handleIFrameLoad,
    hasLoaded,
    iframeHeight,
    showError,
  } = hooks.useIFrameBehavior({
    elementId,
    id,
    iframeUrl,
    onLoaded,
  });

  const {
    modalOptions,
    handleModalClose,
  } = hooks.useModalIFrameData();

  const contentIFrameProps = {
    id: elementId,
    src: iframeUrl,
    allow: IFRAME_FEATURE_POLICY,
    allowFullScreen: true,
    height: iframeHeight,
    scrolling: 'no',
    referrerPolicy: 'origin',
    onLoad: handleIFrameLoad,
  };

  let modalContent;
  if (modalOptions.isOpen) {
    modalContent = modalOptions.body
      ? <div className="unit-modal">{ modalOptions.body }</div>
      : (
        <iframe
          title={modalOptions.title}
          allow={IFRAME_FEATURE_POLICY}
          frameBorder="0"
          src={modalOptions.url}
          style={{ width: '100%', height: modalOptions.height }}
        />
      );
  }

  return (
    <>
      <IFrameComponent
        contentIFrameProps={contentIFrameProps}
        showError={showError}
        title={title}
        hasLoaded={hasLoaded}
        url={iframeUrl}
        loadingMessage={loadingMessage}
        shouldShowContent={shouldShowContent}
      />
      {modalOptions.isOpen && (modalOptions.isFullscreen
        ? (
          <ModalDialog
            dialogClassName="modal-lti"
            onClose={handleModalClose}
            size="fullscreen"
            isOpen
            hasCloseButton={false}
          >
            <ModalDialog.Body className={modalOptions.modalBodyClassName}>
              {modalContent}
            </ModalDialog.Body>
          </ModalDialog>

        ) : (
          <Modal
            body={modalContent}
            dialogClassName="modal-lti"
            onClose={handleModalClose}
            open
          />
        )
      )}
    </>
  );
};

ContentIFrame.propTypes = {
  iframeUrl: PropTypes.string,
  id: PropTypes.string.isRequired,
  shouldShowContent: PropTypes.bool.isRequired,
  loadingMessage: PropTypes.node.isRequired,
  elementId: PropTypes.string.isRequired,
  onLoaded: PropTypes.func,
  title: PropTypes.node.isRequired,
};

ContentIFrame.defaultProps = {
  iframeUrl: null,
  onLoaded: () => ({}),
};

export default ContentIFrame;
