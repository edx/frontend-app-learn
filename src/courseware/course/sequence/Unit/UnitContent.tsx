import React from "react";
import { useUnitContents } from "./data/apiHooks";
import { Spinner } from "@openedx/paragon";
import { AutoXBlock } from "../../../component/XBlock";
import { XBlockRenderContextProvider } from "../../../component/XBlockRenderContext";


interface Props {
    unitId: string;
}

export const UnitContent: React.FC<Props> = ({ unitId, ...props }) => {
    const {data: unitContents, isError, isLoading} = useUnitContents(unitId);

    if (isError) {
        return <p><strong>Error:</strong> Unable to load unit contents.</p>
    }

    if (isLoading || !unitContents) {
        return (
            <Spinner
                animation="border"
                size="xl"
                screenReaderText={"loading"}
            />
        );
    }

    return <main>
        <XBlockRenderContextProvider>
            {unitContents.blocks.map((blockData) => (
                <AutoXBlock key={blockData.id} {...blockData} />
            ))}
        </XBlockRenderContextProvider>
    </main>;
};

export default UnitContent;
