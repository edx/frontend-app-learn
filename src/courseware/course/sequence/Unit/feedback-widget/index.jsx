import React from 'react';
import PropTypes from 'prop-types';
import { useIntl } from '@edx/frontend-platform/i18n';
import { ActionRow, IconButton, Icon } from '@edx/paragon';
import { Close, ThumbUpOutline, ThumbDownOffAlt } from '@edx/paragon/icons';

import './index.scss';
import messages from './messages';
import useFeedbackWidget from './useFeedbackWidget';

const FeedbackWidget = ({
  courseId,
  languageCode,
  unitId,
  userId,
}) => {
  const { formatMessage } = useIntl();
  const {
    closeFeedbackWidget,
    sendFeedback,
    showFeedbackWidget,
    showGratitudeText,
  } = useFeedbackWidget({
    courseId,
    languageCode,
    unitId,
    userId,
  });
  return (
    (showFeedbackWidget || showGratitudeText) && (
      <div className="sequence w-100">
        {
          showFeedbackWidget && (
            <div className="ml-4 mr-2">
              <ActionRow>
                {formatMessage(messages.rateTranslationText)}
                <ActionRow.Spacer />
                <div>
                  <IconButton
                    src={ThumbUpOutline}
                    iconAs={Icon}
                    alt="positive-feedback"
                    onClick={sendFeedback}
                    variant="secondary"
                    className="m-1"
                    id="positive-feedback-button"
                  />
                  <IconButton
                    src={ThumbDownOffAlt}
                    iconAs={Icon}
                    alt="negative-feedback"
                    onClick={sendFeedback}
                    variant="secondary"
                    className="mr-2"
                    id="negative-feedback-button"
                  />
                </div>
                <div className="mb-1 text-light action-row-divider">
                  |
                </div>
                <div>
                  <IconButton
                    src={Close}
                    iconAs={Icon}
                    alt="close-feedback"
                    onClick={closeFeedbackWidget}
                    variant="secondary"
                    className="ml-1 mr-2 float-right"
                    id="close-feedback-button"
                  />
                </div>
              </ActionRow>
            </div>
          )
        }
        {
          showGratitudeText && (
            <div className="ml-4 mr-4">
              <ActionRow className="m-2 justify-content-center">
                {formatMessage(messages.gratitudeText)}
              </ActionRow>
            </div>
          )
        }
      </div>
    )
  );
};

FeedbackWidget.propTypes = {
  courseId: PropTypes.string.isRequired,
  languageCode: PropTypes.string.isRequired,
  userId: PropTypes.string.isRequired,
  unitId: PropTypes.string.isRequired,
};

FeedbackWidget.defaultProps = {};

export default FeedbackWidget;
