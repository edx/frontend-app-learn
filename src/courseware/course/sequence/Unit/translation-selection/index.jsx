import React from 'react';
import PropTypes from 'prop-types';
import { IconButton, Icon, ProductTour } from '@edx/paragon';
import { Language } from '@edx/paragon/icons';
import TranslationModal from './TranslationModal';
import useTranslationTour from './useTranslationTour';

const TranslationSelection = (props) => {
  const {
    translationTour, isOpen, open, close,
  } = useTranslationTour();

  return (
    <>
      <ProductTour tours={[translationTour]} />
      <IconButton
        src={Language}
        iconAs={Icon}
        alt="change-language"
        onClick={open}
        variant="primary"
        className="mr-2 mb-2 float-right"
        id="translation-selection-button"
      />
      <TranslationModal isOpen={isOpen} close={close} {...props} />
    </>
  );
};

TranslationSelection.propTypes = {
  courseId: PropTypes.string.isRequired,
  selectedLanguage: PropTypes.string.isRequired,
  setSelectedLanguage: PropTypes.func.isRequired,
};

TranslationSelection.defaultProps = {};

export default TranslationSelection;
