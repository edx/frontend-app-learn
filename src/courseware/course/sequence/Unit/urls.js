import { getConfig } from '@edx/frontend-platform';
import { stringify } from 'query-string';

export const iframeParams = {
  show_title: 0,
  show_bookmark: 0,
  recheck_access: 1,
};

export const getIFrameUrl = ({
  id,
  view,
  format,
  examAccess,
  translateLanguage,
}) => {
  const xblockUrl = `${getConfig().LMS_BASE_URL}/xblock/${id}`;
  const params = stringify({
    ...iframeParams,
    view,
    ...(format && { format }),
    ...(!examAccess.blockAccess && { exam_access: examAccess.accessToken }),
    ...(translateLanguage && { translate_lang: translateLanguage }),
  });
  return `${xblockUrl}?${params}`;
};

export default {
  getIFrameUrl,
};
