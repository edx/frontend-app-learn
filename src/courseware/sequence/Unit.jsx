import React, { useRef, useEffect, useState } from 'react';
import PropTypes from 'prop-types';
import { getConfig } from '@edx/frontend-platform';
import BookmarkButton from '../bookmark/BookmarkButton';
import { useModel } from '../../model-store';

export default function Unit({
  onLoaded,
  id,
}) {
  const iframeRef = useRef(null);
  const iframeUrl = `${getConfig().LMS_BASE_URL}/xblock/${id}?show_title=0&show_bookmark_button=0`;

  const [iframeHeight, setIframeHeight] = useState(0);
  const [hasLoaded, setHasLoaded] = useState(false);

  const unit = useModel('units', id);

  useEffect(() => {
    global.onmessage = (event) => {
      const { type, payload } = event.data;

      if (type === 'plugin.resize') {
        setIframeHeight(payload.height);
        if (!hasLoaded && iframeHeight === 0 && payload.height > 0) {
          setHasLoaded(true);
          if (onLoaded) {
            onLoaded();
          }
        }
      }
    };
  }, []);

  return (
    <div className="unit">
      <h2 className="mb-0 h4">{unit.title}</h2>
      <BookmarkButton
        unitId={unit.id}
        isBookmarked={unit.bookmarked}
        isProcessing={unit.bookmarkedUpdateState === 'loading'}
      />
      <div className="unit-iframe-wrapper">
        <iframe
          id="unit-iframe"
          title={unit.title}
          ref={iframeRef}
          src={iframeUrl}
          allowFullScreen
          height={iframeHeight}
          scrolling="no"
          referrerPolicy="origin"
        />
      </div>
    </div>
  );
}

Unit.propTypes = {
  id: PropTypes.string.isRequired,
  onLoaded: PropTypes.func,
};

Unit.defaultProps = {
  onLoaded: undefined,
};
