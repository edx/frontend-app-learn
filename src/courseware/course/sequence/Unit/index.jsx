import PropTypes from 'prop-types';
import React from 'react';

import { useIntl } from '@edx/frontend-platform/i18n';

import { useModel } from '@src/generic/model-store';

import BookmarkButton from '../../bookmark/BookmarkButton';
import messages from '../messages';
import UnitContent from './UnitContent';
import UnitSuspense from './UnitSuspense';
import { modelKeys } from './constants';
import { useExamAccess, useShouldDisplayHonorCode } from './hooks';
import UnitTitleSlot from '../../../../plugin-slots/UnitTitleSlot';

const Unit = ({
  courseId,
  format,
  onLoaded,
  id,
}) => {
  const { formatMessage } = useIntl();
  const examAccess = useExamAccess({ id });
  const shouldDisplayHonorCode = useShouldDisplayHonorCode({ courseId, id });
  const unit = useModel(modelKeys.units, id);
  const isProcessing = unit.bookmarkedUpdateState === 'loading';

  const shouldShowContent = !shouldDisplayHonorCode && !examAccess.blockAccess;

  return (
    <div className="unit">
      <div className="mb-0">
        <h3 className="h3">{unit.title}</h3>
        <UnitTitleSlot courseId={courseId} unitId={id} unitTitle={unit.title} />
      </div>
      <h2 className="sr-only">{formatMessage(messages.headerPlaceholder)}</h2>
      <BookmarkButton
        unitId={unit.id}
        isBookmarked={unit.bookmarked}
        isProcessing={isProcessing}
      />
      <UnitSuspense {...{ courseId, id }} />
      <br />
      { shouldShowContent ? <UnitContent unitId={id} /> : null }
    </div>
  );
};

Unit.propTypes = {
  courseId: PropTypes.string.isRequired,
  format: PropTypes.string,
  id: PropTypes.string.isRequired,
  onLoaded: PropTypes.func,
};

Unit.defaultProps = {
  format: null,
  onLoaded: undefined,
};

export default Unit;
