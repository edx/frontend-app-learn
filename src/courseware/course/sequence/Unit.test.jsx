import React from 'react';
// eslint-disable-next-line import/no-extraneous-dependencies
import { cloneDeep } from 'lodash';
import { waitFor } from '@testing-library/dom';
import {
  initialState, messageEvent, render, screen,
} from '../../../setupTest';
import Unit from './Unit';

describe('Unit', () => {
  const mockData = {
    id: '3',
    courseId: '1',
    intl: {},
  };

  it('renders correctly', () => {
    const { asFragment } = render(<Unit {...mockData} />, { initialState });

    expect(screen.getByText('Loading learning sequence...')).toBeInTheDocument();
    expect(screen.getByTitle(mockData.id)).toHaveAttribute('height', String(0));
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders proper message for gated content', () => {
    // Clone initialState.
    const testState = cloneDeep(initialState);
    testState.models.units[mockData.id].graded = true;
    const { asFragment } = render(<Unit {...mockData} />, { initialState: testState });

    expect(screen.getByText('Loading locked content messaging...')).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  it('handles receiving MessageEvent', async () => {
    const { asFragment } = render(<Unit {...mockData} />, { initialState });
    const beforePostingMessage = asFragment();

    window.postMessage(messageEvent, '*');
    // Loading message is gone now.
    await waitFor(() => expect(screen.queryByText('Loading learning sequence...')).not.toBeInTheDocument());
    // Iframe's height is set via message.
    expect(screen.getByTitle(mockData.id)).toHaveAttribute('height', String(messageEvent.payload.height));
    expect(beforePostingMessage).toMatchDiffSnapshot(asFragment());
  });

  it('handles onLoaded after receiving MessageEvent', async () => {
    const onLoaded = jest.fn();
    render(<Unit {...mockData} {...{ onLoaded }} />, { initialState });

    window.postMessage(messageEvent, '*');
    await waitFor(() => expect(onLoaded).toHaveBeenCalledTimes(1));
  });

  it('resizes iframe on second MessageEvent, does not call onLoaded again', async () => {
    const onLoaded = jest.fn();
    // Clone message and set different height.
    const testMessageWithOtherHeight = { ...messageEvent, payload: { height: 200 } };
    render(<Unit {...mockData} {...{ onLoaded }} />, { initialState });

    window.postMessage(messageEvent, '*');
    await waitFor(() => expect(screen.getByTitle(mockData.id)).toHaveAttribute('height', String(messageEvent.payload.height)));
    window.postMessage(testMessageWithOtherHeight, '*');
    await waitFor(() => expect(screen.getByTitle(mockData.id)).toHaveAttribute('height', String(testMessageWithOtherHeight.payload.height)));
    expect(onLoaded).toHaveBeenCalledTimes(1);
  });

  it('ignores MessageEvent with unhandled type', async () => {
    // Clone message and set different type.
    const testMessageWithUnhandledType = { ...messageEvent, type: 'wrong type' };
    render(<Unit {...mockData} />, { initialState });

    window.postMessage(testMessageWithUnhandledType, '*');
    // HACK: We don't have a function we could reliably await here, so this test relies on the timeout of `waitFor`.
    await expect(waitFor(
      () => expect(screen.getByTitle(mockData.id)).toHaveAttribute('height', String(testMessageWithUnhandledType.payload.height)),
      { timeout: 100 },
    )).rejects.toThrowErrorMatchingSnapshot();
  });
});
