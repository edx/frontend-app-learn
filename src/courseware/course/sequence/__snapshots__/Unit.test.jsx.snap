// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Unit handles receiving MessageEvent 1`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -28,33 +28,16 @@
          <span>
            Bookmark this page
          </span>
        </span>
      </button>
-     <div>
      <div
-         class=\\"d-flex justify-content-center align-items-center flex-column\\"
-         style=\\"height: 50vh;\\"
-       >
-         <div
-           class=\\"spinner-border text-primary\\"
-           role=\\"status\\"
-         >
-           <span
-             class=\\"sr-only\\"
-           >
-             Loading learning sequence...
-           </span>
-         </div>
-       </div>
-     </div>
-     <div
        class=\\"unit-iframe-wrapper\\"
      >
        <iframe
          allowfullscreen=\\"\\"
-         height=\\"0\\"
+         height=\\"300\\"
          id=\\"unit-iframe\\"
          referrerpolicy=\\"origin\\"
          scrolling=\\"no\\"
          src=\\"http://localhost:18000/xblock/3?show_title=0&show_bookmark_button=0\\"
          title=\\"3\\""
`;

exports[`Unit ignores MessageEvent with unhandled type 1`] = `
"[2mexpect([22m[31melement[39m[2m).toHaveAttribute([22m[32m[32m\\"height\\"[39m[32m[39m[2m, [22m[32m[32m\\"300\\"[39m[32m[39m[2m) // element.getAttribute(\\"height\\") === \\"300\\"[22m

Expected the element to have attribute:
[32m  height=\\"300\\"[39m
Received:
[31m  height=\\"0\\"[39m"
`;

exports[`Unit renders correctly 1`] = `
<DocumentFragment>
  <div
    class="unit"
  >
    <h2
      class="mb-0 h4"
    >
      3
    </h2>
    <button
      aria-disabled="false"
      aria-live="assertive"
      class="btn pgn__stateful-btn pgn__stateful-btn-state-default btn-link px-1 ml-n1 btn-sm"
      type="button"
    >
      <span
        class="d-flex align-items-center justify-content-center"
      >
        <span
          class="pgn__stateful-btn-icon"
        >
          <img
            alt="fa-bookmark"
            class="fa-bookmark"
            data-testid="icon"
          />
        </span>
        <span>
          Bookmark this page
        </span>
      </span>
    </button>
    <div>
      <div
        class="d-flex justify-content-center align-items-center flex-column"
        style="height: 50vh;"
      >
        <div
          class="spinner-border text-primary"
          role="status"
        >
          <span
            class="sr-only"
          >
            Loading learning sequence...
          </span>
        </div>
      </div>
    </div>
    <div
      class="unit-iframe-wrapper"
    >
      <iframe
        allowfullscreen=""
        height="0"
        id="unit-iframe"
        referrerpolicy="origin"
        scrolling="no"
        src="http://localhost:18000/xblock/3?show_title=0&show_bookmark_button=0"
        title="3"
      />
    </div>
  </div>
</DocumentFragment>
`;

exports[`Unit renders proper message for gated content 1`] = `
<DocumentFragment>
  <div
    class="unit"
  >
    <h2
      class="mb-0 h4"
    >
      3
    </h2>
    <button
      aria-disabled="false"
      aria-live="assertive"
      class="btn pgn__stateful-btn pgn__stateful-btn-state-default btn-link px-1 ml-n1 btn-sm"
      type="button"
    >
      <span
        class="d-flex align-items-center justify-content-center"
      >
        <span
          class="pgn__stateful-btn-icon"
        >
          <img
            alt="fa-bookmark"
            class="fa-bookmark"
            data-testid="icon"
          />
        </span>
        <span>
          Bookmark this page
        </span>
      </span>
    </button>
    <div>
      <div
        class="d-flex justify-content-center align-items-center flex-column"
        style="height: 50vh;"
      >
        <div
          class="spinner-border text-primary"
          role="status"
        >
          <span
            class="sr-only"
          >
            Loading locked content messaging...
          </span>
        </div>
      </div>
    </div>
    <div>
      <div
        class="d-flex justify-content-center align-items-center flex-column"
        style="height: 50vh;"
      >
        <div
          class="spinner-border text-primary"
          role="status"
        >
          <span
            class="sr-only"
          >
            Loading learning sequence...
          </span>
        </div>
      </div>
    </div>
    <div
      class="unit-iframe-wrapper"
    >
      <iframe
        allowfullscreen=""
        height="0"
        id="unit-iframe"
        referrerpolicy="origin"
        scrolling="no"
        src="http://localhost:18000/xblock/3?show_title=0&show_bookmark_button=0"
        title="3"
      />
    </div>
  </div>
</DocumentFragment>
`;
